homeassistant/components/esphome/valve.py

```diff
@@ -0,0 +1,110 @@
+"""Support for ESPHome valves."""
+
+from __future__ import annotations
+
+from typing import Any
+
+from aioesphomeapi import EntityInfo, ValveInfo, ValveOperation, ValveState
+
+from homeassistant.components.valve import (
+    ValveDeviceClass,
+    ValveEntity,
+    ValveEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util.enum import try_parse_enum
+
+from .entity import (
+    EsphomeEntity,
+    convert_api_error_ha_error,
+    esphome_state_property,
+    platform_async_setup_entry,
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up ESPHome valves based on a config entry."""
+    await platform_async_setup_entry(
+        hass,
+        entry,
+        async_add_entities,
+        info_type=ValveInfo,
+        entity_type=EsphomeValve,
+        state_type=ValveState,
+    )
+
+
+class EsphomeValve(EsphomeEntity[ValveInfo, ValveState], ValveEntity):
+    """A valve implementation for ESPHome."""
+
+    @callback
+    def _on_static_info_update(self, static_info: EntityInfo) -> None:
+        """Set attrs from static info."""
+        super()._on_static_info_update(static_info)
+        static_info = self._static_info
+        flags = ValveEntityFeature.OPEN | ValveEntityFeature.CLOSE
+        if static_info.supports_stop:
+            flags |= ValveEntityFeature.STOP
+        if static_info.supports_position:
+            flags |= ValveEntityFeature.SET_POSITION
+        self._attr_supported_features = flags
+        self._attr_device_class = try_parse_enum(
+            ValveDeviceClass, static_info.device_class
+        )
+        self._attr_assumed_state = static_info.assumed_state
+
+    @property
+    @esphome_state_property
+    def is_closed(self) -> bool:
+        """Return if the valve is closed or not."""
+        return self._state.position == 0.0
+
+    @property
+    @esphome_state_property
+    def is_opening(self) -> bool:
+        """Return if the valve is opening or not."""
+        return self._state.current_operation is ValveOperation.IS_OPENING
+
+    @property
+    @esphome_state_property
+    def is_closing(self) -> bool:
+        """Return if the valve is closing or not."""
+        return self._state.current_operation is ValveOperation.IS_CLOSING
+
+    @property
+    @esphome_state_property
+    def reports_position(self) -> bool:
+        """Return True if entity reports position, False otherwise."""
+        return self._static_info.supports_position
```

Comment:
This could be set in _on_static_info_update

Reply:
Thank you! 🍻